File: src/scala/FibonacciMemoization.scala
object FibonacciMemoization {
  import scala.collection.mutable.Map

  val memo: Map[Int, BigInt] = Map()

  def fibonacci(n: Int): BigInt = {
    if (n <= 1) {
      return n
    }

    memo.getOrElseUpdate(n, fibonacci(n - 1) + fibonacci(n - 2))
  }

  def main(args: Array[String]): Unit = {
    val index = 15
    println(s"Fibonacci (memoization) of $index is: ${fibonacci(index)}")
  }
}

File: src/rust/fibonacci_memoization.rs
use std::collections::HashMap;

fn fibonacci(n: u32, memo: &mut HashMap<u32, u64>) -> u64 {
    if n <= 1 {
        return n as u64;
    }

    // Check if the result is already memoized
    if let Some(&cached_result) = memo.get(&n) {
        return cached_result;
    }

    // Calculate Fibonacci recursively and store the result in memoization
    let result = fibonacci(n - 1, memo) + fibonacci(n - 2, memo);
    memo.insert(n, result);

    result
}

fn main() {
    let index = 15;
    let mut memo: HashMap<u32, u64> = HashMap::new();

    println!(
        "Fibonacci (memoization) of {} is: {}",
        index,
        fibonacci(index, &mut memo)
    );
}
